### PR TITLE
[FLINK-12807][hive]Support Hive table columnstats related operations in HiveCatalog

### PR DESCRIPTION
**Pull request by zjuwangg**
## What is the purpose of the change

* Support Hive table column stats related operations in HiveCatalog

## Brief change log

  - *support Hive table column stats related operations in HiveCatalog, such as to get table/partition column stats and alter column stats.*

## Verifying this change
This change added tests and can be verified as follows:

*(example:)*
  - *Added test that validates that column stats related operation can be done expectedly*

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): (no)
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: (no)
  - The serializers: (no)
  - The runtime per-record code paths (performance sensitive): (no)
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Yarn/Mesos, ZooKeeper: (no)
  - The S3 file system connector: (no)

## Documentation

  - Does this pull request introduce a new feature? (no)
  - If yes, how is the feature documented? (not documented)

### CODE REVIEW

**Comment by zjuwangg** [comment]
cc @bowenli86 @xuefuz @lirui-apache  to review.

File: flink-connectors/flink-connector-hive/src/main/java/org/apache/flink/table/catalog/hive/util/HiveCatalogUtil.java [hunk 1]
```diff
@@ -0,0 +1,235 @@
...
+				return new CatalogColumnStatisticsDataDouble(
+						doubleStats.getLowValue(), doubleStats.getHighValue(),
+						doubleStats.getNumDVs(), doubleStats.getNumNulls());
+		} else if (stats.isSetLongStats()) {
+				LongColumnStatsData longColStats = stats.getLongStats();
+				return new CatalogColumnStatisticsDataLong(
+						longColStats.getLowValue(), longColStats.getHighValue(),
+						longColStats.getNumDVs(), longColStats.getNumNulls());
+		} else if (stats.isSetStringStats()) {
+			StringColumnStatsData stringStats = stats.getStringStats();
```
Comment (bowenli86) [inline flink-connectors/flink-connector-hive/src/main/java/org/apache/flink/table/catalog/hive/util/HiveCatalogUtil.java]: reformat the indent?

File: flink-connectors/flink-connector-hive/src/main/java/org/apache/flink/table/catalog/hive/util/HiveCatalogUtil.java [hunk 2]
```diff
@@ -0,0 +1,235 @@
...
+		} else if (stats.isSetDateStats()) {
+			DateColumnStatsData dateStats = stats.getDateStats();
+			return new CatalogColumnStatisticsDataDate(
+					new org.apache.flink.table.catalog.stats.Date(dateStats.getLowValue().getDaysSinceEpoch()),
+					new org.apache.flink.table.catalog.stats.Date(dateStats.getHighValue().getDaysSinceEpoch()),
+					dateStats.getNumDVs(),
+					dateStats.getNumNulls());
+		} else if (stats.isSetDoubleStats()) {
+				DoubleColumnStatsData doubleStats = stats.getDoubleStats();
+				return new CatalogColumnStatisticsDataDouble(
```
Comment (bowenli86) [inline flink-connectors/flink-connector-hive/src/main/java/org/apache/flink/table/catalog/hive/util/HiveCatalogUtil.java]: minor: to improve readability, how about reformat it to 

```
return new CatalogColumnStatisticsDataDouble(
						doubleStats.getLowValue(),
						doubleStats.getHighValue(),
						doubleStats.getNumDVs(),
						doubleStats.getNumNulls());
```

Same to code below

File: flink-connectors/flink-connector-hive/src/main/java/org/apache/flink/table/catalog/hive/util/HiveCatalogUtil.java [hunk 3]
```diff
@@ -0,0 +1,235 @@
...
+			StringColumnStatsData stringStats = stats.getStringStats();
+			return new CatalogColumnStatisticsDataString(
+					stringStats.getMaxColLen(), stringStats.getAvgColLen(),
+					stringStats.getNumDVs(), stringStats.getNumNulls());
+		} else {
+			LOG.warn("Flink does not support converting ColumnStatisticsData '{}' for Hive column type '{}' yet.", stats, colType);
+			return null;
+		}
+	}
+
```
Comment (bowenli86) [inline flink-connectors/flink-connector-hive/src/main/java/org/apache/flink/table/catalog/hive/util/HiveCatalogUtil.java]: remove extra line

File: flink-connectors/flink-connector-hive/src/main/java/org/apache/flink/table/catalog/hive/util/HiveCatalogUtil.java [hunk 4]
```diff
@@ -0,0 +1,235 @@
...
+
+	/**
+	 * Convert Flink ColumnStats to Hive ColumnStatisticsData according to Hive column type.
+	 * Note we currently assume that, in Flink, the max and min of ColumnStats will be same type as the Flink column type.
+	 * For example, for SHORT and Long columns, the max and min of their ColumnStats should be of type SHORT and LONG.
+	 */
+	private static ColumnStatisticsData getColumnStatisticsData(DataType colType, CatalogColumnStatisticsDataBase colStat) {
+		LogicalType colLogicalType = colType.getLogicalType();
```
Comment (bowenli86) [inline flink-connectors/flink-connector-hive/src/main/java/org/apache/flink/table/catalog/hive/util/HiveCatalogUtil.java]: use logical type root for comparison. Please reference `HiveTypeUtil`.

File: flink-connectors/flink-connector-hive/src/main/java/org/apache/flink/table/catalog/hive/util/HiveCatalogUtil.java [hunk 5]
```diff
@@ -0,0 +1,235 @@
...
+		if (colLogicalType instanceof CharType || colLogicalType instanceof VarCharType) {
```
Comment (bowenli86) [inline flink-connectors/flink-connector-hive/src/main/java/org/apache/flink/table/catalog/hive/util/HiveCatalogUtil.java]: `StringType` too?

File: flink-connectors/flink-connector-hive/src/main/java/org/apache/flink/table/catalog/hive/util/HiveCatalogUtil.java [hunk 6]
```diff
@@ -0,0 +1,235 @@
...
+		} else if (colLogicalType instanceof BooleanType) {
+			if (colStat instanceof CatalogColumnStatisticsDataBoolean) {
+				CatalogColumnStatisticsDataBoolean booleanColStat = (CatalogColumnStatisticsDataBoolean) colStat;
+				BooleanColumnStatsData boolStats = new BooleanColumnStatsData(
+						booleanColStat.getTrueCount(),
+						booleanColStat.getFalseCount(),
+						booleanColStat.getNullCount());
+				return ColumnStatisticsData.booleanStats(boolStats);
+			}
+		} else if (colLogicalType instanceof TinyIntType || colLogicalType instanceof SmallIntType ||
```
Comment (bowenli86) [inline flink-connectors/flink-connector-hive/src/main/java/org/apache/flink/table/catalog/hive/util/HiveCatalogUtil.java]: minor: to improve readability, I'd reformat it to:

```
} else if (colLogicalType instanceof TinyIntType 
    || colLogicalType instanceof SmallIntType 
    || colLogicalType instanceof IntType 
    || colLogicalType instanceof BigIntType 
    || colLogicalType instanceof TimestampType) {
  ...
}
```

File: flink-connectors/flink-connector-hive/src/main/java/org/apache/flink/table/catalog/hive/util/HiveCatalogUtil.java [hunk 7]
```diff
@@ -0,0 +1,235 @@
...
+		} else if (colLogicalType instanceof DateType) {
+			if (colStat instanceof CatalogColumnStatisticsDataDate) {
+				CatalogColumnStatisticsDataDate dateColumnStatsData = (CatalogColumnStatisticsDataDate) colStat;
+				DateColumnStatsData dateStats = new DateColumnStatsData(dateColumnStatsData.getNullCount(), dateColumnStatsData.getNdv());
+				dateStats.setHighValue(new org.apache.hadoop.hive.metastore.api.Date(dateColumnStatsData.getMax().getDaysSinceEpoch()));
+				dateStats.setLowValue(new org.apache.hadoop.hive.metastore.api.Date(dateColumnStatsData.getMin().getDaysSinceEpoch()));
+				return ColumnStatisticsData.dateStats(dateStats);
+			}
+		}
+		LOG.warn("Flink does not support converting ColumnStats '{}' for Hive column type '{}' yet.", colStat, colType);
```
Comment (bowenli86) [inline flink-connectors/flink-connector-hive/src/main/java/org/apache/flink/table/catalog/hive/util/HiveCatalogUtil.java]: throw exception?

File: flink-connectors/flink-connector-hive/src/test/java/org/apache/flink/table/catalog/hive/HiveCatalogHiveMetadataTest.java [hunk 8]
```diff
@@ -70,6 +77,35 @@ public void testCreateTable_StorageFormatSet() throws Exception {
...
+		catalog.createDatabase(db1, createDb(), false);
+		CatalogTable catalogTable = createTable();
+		catalog.createTable(path1, catalogTable, false);
+		Map<String, CatalogColumnStatisticsDataBase> columnStatisticsDataBaseMap = new HashMap<>();
+		columnStatisticsDataBaseMap.put("first", new CatalogColumnStatisticsDataString(10, 5.2, 3, 100));
+		columnStatisticsDataBaseMap.put("second", new CatalogColumnStatisticsDataLong(0, 1000, 3, 0));
+		columnStatisticsDataBaseMap.put("third", new CatalogColumnStatisticsDataString(15, 5.2, 3, 50));
+		CatalogColumnStatistics catalogColumnStatistics = new CatalogColumnStatistics(columnStatisticsDataBaseMap);
+		catalog.alterTableColumnStatistics(path1, catalogColumnStatistics, false);
+		CatalogTestUtil.checkEquals(catalogColumnStatistics, catalog.getTableColumnStatistics(path1));
```
Comment (bowenli86) [inline flink-connectors/flink-connector-hive/src/test/java/org/apache/flink/table/catalog/hive/HiveCatalogHiveMetadataTest.java]: nit: add an empty line to separate logic code and assertions to improve readability?

File: flink-connectors/flink-connector-hive/src/test/java/org/apache/flink/table/catalog/hive/HiveCatalogHiveMetadataTest.java [hunk 9]
```diff
@@ -70,6 +77,35 @@ public void testCreateTable_StorageFormatSet() throws Exception {
 		assertFalse(StringUtils.isNullOrWhitespaceOnly(serde));
 	}
 
+	// ------ table and column stats ------
+	@Test
+	public void testAlter_GET_TableColumnStatistics() throws Exception {
```
Comment (bowenli86) [inline flink-connectors/flink-connector-hive/src/test/java/org/apache/flink/table/catalog/hive/HiveCatalogHiveMetadataTest.java]: can we follow the naming conventions? This should probably just be `testAlterTableColumnStatistics()`

File: flink-connectors/flink-connector-hive/src/main/java/org/apache/flink/table/catalog/hive/util/HiveCatalogUtil.java [hunk 10]
```diff
@@ -0,0 +1,235 @@
...
+
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+
+/**
+ * Utils to for HiveCatalog.
+ */
+public class HiveCatalogUtil {
```
Comment (bowenli86) [inline flink-connectors/flink-connector-hive/src/main/java/org/apache/flink/table/catalog/hive/util/HiveCatalogUtil.java]: rename to `HiveStatsUtil`?

File: flink-table/flink-table-common/src/test/java/org/apache/flink/table/catalog/CatalogTestUtil.java [hunk 11]
```diff
@@ -58,7 +58,7 @@ static void checkEquals(CatalogTableStatistics ts1, CatalogTableStatistics ts2)
 		assertEquals(ts1.getProperties(), ts2.getProperties());
 	}
 
-	static void checkEquals(CatalogColumnStatistics cs1, CatalogColumnStatistics cs2) {
+	public static void checkEquals(CatalogColumnStatistics cs1, CatalogColumnStatistics cs2) {
```
Comment (bowenli86) [inline flink-table/flink-table-common/src/test/java/org/apache/flink/table/catalog/CatalogTestUtil.java]: move this to CatalogTestBase?

File: flink-connectors/flink-connector-hive/src/main/java/org/apache/flink/table/catalog/hive/HiveCatalog.java [hunk 12]
```diff
@@ -1069,7 +1075,22 @@ public void alterTableStatistics(ObjectPath tablePath, CatalogTableStatistics ta
...
 	@Override
 	public void alterTableColumnStatistics(ObjectPath tablePath, CatalogColumnStatistics columnStatistics, boolean ignoreIfNotExists) throws TableNotExistException, CatalogException {
-
+		try {
+			Table hiveTable = getHiveTable(tablePath);
+			// Set table column stats. This only works for non-partitioned tables.
+			if (!isTablePartitioned(hiveTable)) {
+				client.updateTableColumnStatistics(HiveCatalogUtil.createTableColumnStats(hiveTable, columnStatistics.getColumnStatisticsData()));
+			} else {
+				throw new CatalogException(String.format("Failed to alter partition table column stats of table %s",
```
[the remark this answers is not in the document]
Comment (xuefuz) [reply flink-connectors/flink-connector-hive/src/main/java/org/apache/flink/table/catalog/hive/HiveCatalog.java]: Yeah. I think TableNotPartitioned makes sense.

[the remark this answers is not in the document]
Comment (zjuwangg) [reply flink-connectors/flink-connector-hive/src/main/java/org/apache/flink/table/catalog/hive/HiveCatalog.java]: exactly

File: flink-connectors/flink-connector-hive/src/main/java/org/apache/flink/table/catalog/hive/HiveCatalog.java [hunk 13]
```diff
@@ -1079,7 +1100,27 @@ public void alterPartitionStatistics(ObjectPath tablePath, CatalogPartitionSpec
 
 	@Override
 	public void alterPartitionColumnStatistics(ObjectPath tablePath, CatalogPartitionSpec partitionSpec, CatalogColumnStatistics columnStatistics, boolean ignoreIfNotExists) throws PartitionNotExistException, CatalogException {
+		try {
+			Partition hivePartition = getHivePartition(tablePath, partitionSpec);
+			Table hiveTable = getHiveTable(tablePath);
+			String partName = getPartitionName(tablePath, partitionSpec, hiveTable);
```
[the remark this answers is not in the document]
Comment (zjuwangg) [reply flink-connectors/flink-connector-hive/src/main/java/org/apache/flink/table/catalog/hive/HiveCatalog.java]: Yes, we can not just obtain partition keys from hive Partition

File: flink-connectors/flink-connector-hive/src/main/java/org/apache/flink/table/catalog/hive/HiveCatalog.java [hunk 14]
```diff
@@ -1074,15 +1074,14 @@ public void alterTableStatistics(ObjectPath tablePath, CatalogTableStatistics ta
...
 		try {
 			Table hiveTable = getHiveTable(tablePath);
 			// Set table column stats. This only works for non-partitioned tables.
 			if (!isTablePartitioned(hiveTable)) {
-				client.updateTableColumnStatistics(HiveCatalogUtil.createTableColumnStats(hiveTable, columnStatistics.getColumnStatisticsData()));
+				client.updateTableColumnStatistics(HiveStatsUtil.createTableColumnStats(hiveTable, columnStatistics.getColumnStatisticsData()));
 			} else {
-				throw new CatalogException(String.format("Failed to alter partition table column stats of table %s",
-														tablePath.getFullName()));
+				throw new TableNotPartitionedException(getName(), tablePath);
```
Comment (bowenli86) [inline flink-connectors/flink-connector-hive/src/main/java/org/apache/flink/table/catalog/hive/HiveCatalog.java]: shouldn't this be a `TablePartitionedException`? not `TableNotPartitionedException`

we need to add a `TablePartitionedException.class`

File: flink-connectors/flink-connector-hive/src/main/java/org/apache/flink/table/catalog/hive/HiveCatalog.java [hunk 15]
```diff
@@ -1116,11 +1117,14 @@ public void alterPartitionColumnStatistics(ObjectPath tablePath, CatalogPartitio
...
-		List<String> partKVs = new ArrayList<>();
+		StringBuilder stringBuilder = new StringBuilder();
 		for (int i = 0; i < partitionCols.size(); i++) {
-			partKVs.add(partitionCols.get(i) + "=" + partitionVals.get(i));
+			stringBuilder.append(partitionCols.get(i)).append("=").append(partitionVals.get(i));
+			if (i < partitionCols.size() - 1) {
+				stringBuilder.append("/");
+			}
 		}
-		return org.apache.commons.lang3.StringUtils.join(partKVs, "/");
```
Comment (bowenli86) [inline flink-connectors/flink-connector-hive/src/main/java/org/apache/flink/table/catalog/hive/HiveCatalog.java]: we can keep the `List<String> partKVs` and just call `return String.join("/", partKVs)` at the end

File: flink-connectors/flink-connector-hive/src/main/java/org/apache/flink/table/catalog/hive/util/HiveStatsUtil.java [hunk 16]
```diff
@@ -162,38 +148,44 @@ private static CatalogColumnStatisticsDataBase createTableColumnStats(DataType c
...
 	 * Convert Flink ColumnStats to Hive ColumnStatisticsData according to Hive column type.
 	 * Note we currently assume that, in Flink, the max and min of ColumnStats will be same type as the Flink column type.
 	 * For example, for SHORT and Long columns, the max and min of their ColumnStats should be of type SHORT and LONG.
 	 */
 	private static ColumnStatisticsData getColumnStatisticsData(DataType colType, CatalogColumnStatisticsDataBase colStat) {
-		LogicalType colLogicalType = colType.getLogicalType();
-		if (colLogicalType instanceof CharType || colLogicalType instanceof VarCharType) {
+		LogicalTypeRoot type = colType.getLogicalType().getTypeRoot();
+		if (type.equals(LogicalTypeRoot.CHAR)
+			|| type.equals(LogicalTypeRoot.VARCHAR)) {
```
Comment (bowenli86) [inline flink-connectors/flink-connector-hive/src/main/java/org/apache/flink/table/catalog/hive/util/HiveStatsUtil.java]: nit: one extra tab here

**Comment by zjuwangg** [comment]
cc @bowenli86 to review again

**Comment by bowenli86** [comment]
@zjuwangg LGTM, thanks for your contribution!

@flinkbot approve all

Merging

File: flink-connectors/flink-connector-hive/src/main/java/org/apache/flink/table/catalog/hive/util/HiveStatsUtil.java [hunk 17]
```diff
@@ -0,0 +1,241 @@
...
+
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+
+import static org.apache.flink.util.Preconditions.checkNotNull;
+
+/**
+ * Utils to for HiveCatalog.
```
Comment (bowenli86) [inline flink-connectors/flink-connector-hive/src/main/java/org/apache/flink/table/catalog/hive/util/HiveStatsUtil.java]: I will help update the javadoc upon merging

Comment (zjuwangg) [reply flink-connectors/flink-connector-hive/src/main/java/org/apache/flink/table/catalog/hive/util/HiveStatsUtil.java]: thanks!

**Pull request closed without merging**